exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,384 @@
+	defaultHistogramBuckets []float64
+}
+
+var _ export.Exporter = &Exporter{}
+var _ http.Handler = &Exporter{}
+var _ export.LabelEncoder = &Exporter{}
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registry is the prometheus registry that will be used as the default Registerer and
+	// Gatherer if these are not specified.
+	//
+	// If not set a new empty Registry is created.
+	Registry *prometheus.Registry
+
+	// Registerer is the prometheus registerer to register
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the system-default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// LabelEncoder is the label encoder that will be used to group
+	// and export metrics to Prometheus.
+	LabelEncoder export.LabelEncoder
+}
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registry == nil {
+		opts.Registry = prometheus.NewRegistry()
+	}
+
+	if opts.Registerer == nil {
+		opts.Registerer = opts.Registry
+	}
+
+	if opts.Gatherer == nil {
+		opts.Gatherer = opts.Registry
+	}
+
+	if opts.LabelEncoder == nil {
+		opts.LabelEncoder = metric.DefaultLabelEncoder()
+	}
+
+	return &Exporter{
+		labelEncoder: opts.LabelEncoder,
+
+		registerer: opts.Registerer,
+		gatherer:   opts.Gatherer,
+		handler:    promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}),
+
+		counters:    make(map[metricKey]prometheus.Counter),
+		counterVecs: make(map[*export.Descriptor]*prometheus.CounterVec),
+
+		gauges:    make(map[metricKey]prometheus.Gauge),
+		gaugeVecs: make(map[*export.Descriptor]*prometheus.GaugeVec),
+
+		histograms:    make(map[metricKey]prometheus.Observer),
+		histogramVecs: make(map[*export.Descriptor]*prometheus.HistogramVec),
+
+		defaultHistogramBuckets: opts.DefaultHistogramBuckets,
+	}, nil
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var forEachError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+
+		desc := record.Descriptor()
+
+		labels := record.Labels()
+		if labels.Encoder() != e {
+			// TODO: This case could be handled by directly
+			// encoding the labels at this point, but presently it
+			// should not occur.
+			//
+			// copy from datadog exporter.
+			panic("Should have self-encoded labels")
+		}
+		mKey := metricKey{
+			desc: desc,
+			// TODO: check if encoder is the exporter.
+			encoded: record.Labels().Encoded(),
+		}
+
+		// TODO(paivagustavo): how to choose between Histogram and Summary?
```

Comment:
I think for now this can be in the exporter config. The user decides between histogram and summary at construction time. 

In the future it would be nice to give a way to configure this on a per metric basis, but I am not very certain how that would looke..

Reply:
Yes, I'm interested in more flexible and configurable exporters. I believe there's a question as to whether such a configuration can be in any way shared by more than one exporter.

Reply:
I've added an option to change how measures are exported, defaulting to histograms.